### PR TITLE
VRP Transceivers Support

### PR DESCRIPTION
**Pull request by kaustubh6199**
Update for transceiver

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [x] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

File: LibreNMS/OS/Vrp.php [hunk 1]
```diff
@@ -95,7 +97,38 @@ public function discoverEntityPhysical(): Collection
 
         return $inventory;
     }
+     public function discoverTransceivers(): Collection
+   {
+    // Get a map of ifIndex to port_id for proper association with ports
+    $ifIndexToPortId = $this->getDevice()->ports()->pluck('port_id', 'ifIndex');
+
+    // Walk through the MIB table for transceiver information
+    return SnmpQuery::cache()->walk('HUAWEI-ENTITY-EXTENT-MIB::hwOpticalModuleInfoTable')->mapTable(function ($data, $ifIndex) use ($ifIndexToPortId) {
```
Comment (murrant) [inline LibreNMS/OS/Vrp.php:106]:
```suggestion
    return SnmpQuery::walk('HUAWEI-ENTITY-EXTENT-MIB::hwOpticalModuleInfoTable')->mapTable(function ($data, $ifIndex) use ($ifIndexToPortId) {
```

File: LibreNMS/OS/Vrp.php [hunk 2]
```diff
@@ -95,7 +97,38 @@ public function discoverEntityPhysical(): Collection
...
+            return null;
+        }
 
+        // Handle cases where required data might not be available (fallback to null)
+        $cable = $data['HUAWEI-ENTITY-EXTENT-MIB::hwEntityOpticalCableType'] ?? null;
+        $distance = $data['HUAWEI-ENTITY-EXTENT-MIB::hwEntityOpticalDistance'] ?? null;
+        $wavelength = $data['HUAWEI-ENTITY-EXTENT-MIB::hwEntityOpticalWaveLengthExact'] ?? null;
+
+        // Create a new Transceiver object with the retrieved data
+        return new Transceiver([
```
Comment (murrant) [inline LibreNMS/OS/Vrp.php:118]: You are missing the import for Transceiver :)

Comment (murrant) [reply LibreNMS/OS/Vrp.php:118]: `use App\Models\Transceiver;`

**Comment by kaustubh6199** [comment]
After implementing suggested changes i am getting error "Error discovering Transceiver Module please check logs"

**Review by kaustubh6199 — COMMENTED** [review]
StyleCI recommended Changes

File: LibreNMS/OS/Vrp.php (hunk 2 above)
Comment (kaustubh6199) [reply LibreNMS/OS/Vrp.php]: Done

**Comment by PipoCanaja** [comment]
Hello @kaustubh6199 
And welcome in the project ! Thanx for your pull request. I have a fair amount of Huawei routers, and this was in my todo list. I'll test your PR in a few devices to see how it goes, and we'll get it ready for merge.

**Comment by PipoCanaja** [comment]
Now we have the transceiver discovered, at least on a 5731: 
image

But No way to display the graphs -> They are indexed by entPhysicalIndex and not ifIndex. And we have no way to do that in the YAML unfortunately cause there is no OID to query that would return ifIndex directly.

**Comment by kaustubh6199** [comment]
> Now we have the transceiver discovered, at least on a 5731: image
> 
> But No way to display the graphs -> They are indexed by entPhysicalIndex and not ifIndex. And we have no way to do that in the YAML unfortunately cause there is no OID to query that would return ifIndex directly.

Can we use the existing dbm graphs for the it

**Comment by kaustubh6199** [comment]
> Now we have the transceiver discovered, at least on a 5731: image
> 
> But No way to display the graphs -> They are indexed by entPhysicalIndex and not ifIndex. And we have no way to do that in the YAML unfortunately cause there is no OID to query that would return ifIndex directly.

i am taking reference from https://support.huawei.com/enterprise/en/doc/EDOC1000178181/b59ea7c9/huawei-entity-extent-mib#dc_s_mib_entity-extent_0007 this and found the exact oids for it can we use it for graph and showing wavelengths ??

**Comment by PipoCanaja** [comment]
We already have the graphs, but we lack a way to link them to the newly discovered "Transceivers". We need to find a solution for that, I am searching currently.

**Comment by PipoCanaja** [comment]
Getting to it (on a CE 12804 here): 
image

**Comment by kaustubh6199** [comment]
Yup its working for S5700 switch also 
Screenshot 2024-10-05 102607
Screenshot 2024-10-05 102627

**Review by kaustubh6199 — COMMENTED** [review]
The Changes made are working for VRP Switch like S5720-32X-EI-24S-AC ,S5735-L8T4X-A1,CE6851HI

File: LibreNMS/OS/Vrp.php [hunk 3]
```diff
@@ -96,6 +97,59 @@ public function discoverEntityPhysical(): Collection
         return $inventory;
     }
 
+    public function discoverTransceivers(): Collection
+    {
+        // Get a map of ifIndex to port_id for proper association with ports
+        $ifIndexToPortId = $this->getDevice()->ports()->pluck('port_id', 'ifIndex');
+
+        // EntityPhysicalIndex to ifIndex
+        $entityToIfIndex = \SnmpQuery::walk('ENTITY-MIB::entAliasMappingIdentifier')->mapTable(function ($data, $entPhysicalIndex) {
```
Comment (murrant) [inline LibreNMS/OS/Vrp.php:106]: The entity-physical module should already be fetching this oid and storing ifIndex. You could save some work by using that, I do something similar in the Cisco code.  Perhaps I will add some common code to use to help with this.

File: LibreNMS/OS/Vrp.php [hunk 4]
```diff
@@ -96,6 +97,59 @@ public function discoverEntityPhysical(): Collection
...
+            $port_id = $ifIndexToPortId->get($ifIndex) ?? null;
+            if (is_null($port_id)) {
+                // Invalid
+                return null;
+            }
+
+            // Create a new Transceiver object with the retrieved data
+            return new Transceiver([
+                'port_id' => $ifIndexToPortId->get($ifIndex),  // Map ifIndex to port_id
+                'index' => $ifIndex,
```
Comment (murrant) [inline LibreNMS/OS/Vrp.php:140]: This should be $entIndex

Comment (murrant) [inline LibreNMS/OS/Vrp.php:140]:
```suggestion
                'index' => $entIndex,
```

This field is for the index within hwOpticalModuleInfoTable

File: LibreNMS/OS/Vrp.php (hunk 3 above)
Comment (murrant) [reply LibreNMS/OS/Vrp.php:106]: This actually exists :)

$entityToIfIndex = $this->getIfIndexEntPhysicalMap();

**Review by kaustubh6199 — COMMENTED** [review]
Need input on this matter from both of you if i am going in right way or not

**Review by kaustubh6199 — COMMENTED** [review]
styleci

File: LibreNMS/OS/Vrp.php (hunk 3 above)
Comment (kaustubh6199) [reply LibreNMS/OS/Vrp.php:106]: Hey @murrant @PipoCanaja where i am getting wrong

Comment (murrant) [reply LibreNMS/OS/Vrp.php:106]: You can replace the SnmpQuery with `$entityToIfIndex = $this->getIfIndexEntPhysicalMap();` which will give you the same map without duplicating an snmpwalk.

**Comment by PipoCanaja** [comment]
Testing on various VRP devices: 
[ ] - S5720-56C-PWR-EI-AC, V200R019C10SPC500
[ ] - S5731-H48P4XC, V200R022C00SPC500
[ ] - CE5855EI, V200R005C10SPC800

**Comment by PipoCanaja** [comment]
Looks good to me. Checking the test-data right now

**Comment by PipoCanaja** [comment]
Tests OK, good step. 

Checking `$this->getIfIndexEntPhysicalMap();` now.

**Comment by PipoCanaja** [comment]
`$this->getIfIndexEntPhysicalMap();`  fixed. No impact on tests. 

LGTM

**Comment by PipoCanaja** [comment]
Well done @kaustubh6199 , thanx for your contribution !

**Review by PipoCanaja — DISMISSED** [review]
Tested OK on various devices. LGTM, ready for merge

**Review by PipoCanaja — APPROVED** [review]
LGTM, including distance :D

**Comment by kaustubh6199** [comment]
I much appreciate your support as you know i am not developer but just an enthusiast for networking .
I have one question what if i want to give description shown instated of interface name then what i have to do in the vrp.php file

**Pull request merged**